### PR TITLE
fix(safety): add path traversal and symlink protection to write operations

### PR DESCRIPTION
**Pull request by junhoyeo**
## Summary

Adds path traversal protection, symlink resolution, and atomic write semantics to all file-mutating tools (`write_file`, `edit_file`, `delete_file`).

## Changes

- **`assertWriteSafety` guard** — new validation function that rejects any resolved path outside the project root. Applied to write, edit, and delete operations.
- **Symlink resolution** — uses `lstat` + `realpath` to detect and block writes through symlinks that escape the project root. Handles macOS `/var` → `/private/var` canonical path difference via separate `resolvedRoot` (path traversal check) and `canonicalRoot` (symlink detection).
- **Atomic writes** — `write_file` now writes to a temp file and renames into place, eliminating the window between check and write.
- **Test coverage** — 115 modify-subsystem tests pass, including new traversal/symlink/TOCTOU edge cases.

## Files Changed

| File | Change |
|------|--------|
| `src/tools/utils/safety-utils.ts` | New `assertWriteSafety`, symlink detection, root resolution |
| `src/tools/modify/write-file.ts` | Atomic write (temp + rename), safety guard |
| `src/tools/modify/edit-file.ts` | Safety guard before mutation |
| `src/tools/modify/delete-file.ts` | Safety guard, symlink-aware unlink |
| `src/tools/modify/*.test.ts` | New test cases for traversal, symlinks, TOCTOU |

### CODE REVIEW

**Comment by junhoyeo** [comment]
Thanks for the thorough reviews! All issues have been addressed in the force-pushed update:

**Root canonicalization** (@cubic-dev-ai, @chatgpt-codex-connector) — `assertWriteSafety` now computes a `canonicalRoot` via `realpath(resolvedRoot)` and passes it to the symlink guard. This handles macOS `/var` → `/private/var` correctly. The path traversal guard still uses `resolvedRoot` (logical containment), while symlink guards use `canonicalRoot` (physical resolution).

**lstat error swallowing** (@cubic-dev-ai) — The catch block in `safeAtomicWriteFile` now only treats `ENOENT` as "file doesn't exist". All other errors (`EACCES`, `EIO`, etc.) are re-thrown instead of silently bypassing the symlink safety check.

**File mode preservation** (@chatgpt-codex-connector) — `safeAtomicWriteFile` now reads the original file's permission bits via `lstat().mode` and applies them to the temp file via `open(tmpPath, "wx", originalMode)`. New files still default to `0o644`.

**Atomic write hardening** — `safeAtomicWriteFile` has been moved from `write-file.ts` into `safety-utils.ts` alongside the other safety functions. Now uses `crypto.randomBytes` for unpredictable temp names and `O_EXCL` flag to prevent symlink pre-creation attacks on the temp path.

**Ancestor symlink detection** (@gemini-code-assist) — The code already walks ancestor directories for non-existent files via `checkAncestorSymlinkSafety`, and checks intermediate directory symlinks for existing files via `realpath()` comparison. Both now use the canonicalized root.

File: packages/cea/src/tools/modify/edit-file.ts [hunk 1]
```diff
@@ -209,8 +218,8 @@ export async function executeEditFile(input: EditFileInput): Promise<string> {
 
   const writeContent = restoreFileText(canonicalNewContent, oldEnvelope);
 
-  await ensureParentDir(parsed.path);
-  await writeFile(parsed.path, writeContent, "utf-8");
+  await ensureParentDir(safePath);
+  await writeFile(safePath, writeContent, "utf-8");
```
[the remark this answers is not in the document]
Comment (junhoyeo) [reply packages/cea/src/tools/modify/edit-file.ts]: The edit-file path runs through `assertWriteSafety` which validates the path before any mutation. The actual file content read+write in edit-file is inherently sequential (read → compute edits → write), so the TOCTOU window here is the same as any filesystem operation. The atomic write pattern is applied to `write_file` where the check-then-create race was most exploitable.

File: packages/cea/src/tools/utils/safety-utils.ts [hunk 2]
```diff
@@ -615,3 +615,207 @@ export async function safeReadFileEnhanced(
...
+ * @throws Error if any safety guard rejects the path
+ */
+export async function assertWriteSafety(
+  filePath: string,
+  rootDir: string = process.cwd()
+): Promise<string> {
+  const resolvedRoot = resolve(rootDir);
+  const resolvedFilePath = isAbsolute(filePath)
+    ? resolve(filePath)
+    : resolve(resolvedRoot, filePath);
```
[the remark this answers is not in the document]
Comment (junhoyeo) [reply packages/cea/src/tools/utils/safety-utils.ts]: Fixed. `assertWriteSafety` now computes `canonicalRoot = await realpath(resolvedRoot)` and passes it to symlink guards. Path traversal guard uses `resolvedRoot` (logical), symlink guards use `canonicalRoot` (physical). Handles macOS `/var` → `/private/var`.

**Review by minpeter — APPROVED** [review]
LGTM
